### ISSUE TITLE
make some normalizing flows demos

### ISSUE THREAD

**Issue by murphyk** [probml/pyprobml#672]
Currently the flows chapter only has [one demo ](https://colab.research.google.com/github/probml/probml-notebooks/blob/master/notebooks/flow_2d_mlp.ipynb) in haiku for modeling a simple 2d dataset, and the rest of the chapter is 'dry' theory. It would be useful to create pedagogical demos to illustrate the use of flows for density estimation, generative modeling, and posterior inference.

Concretely, here are some things to do

1.  take the [distrax  MNIST flow demo ](https://github.com/deepmind/distrax/blob/master/examples/flow.py)
and convert it to a standalone colab (similar in spirit to the demo above).
2. take the [jax-flows two moons demo](https://github.com/ChrisWaites/jax-flows/blob/master/examples/intro.ipynb) and recreate it using distrax and make a colab
3. take some [pzflow examples](https://github.com/jfcrenshaw/pzflow) and apply them to more 'standard' tabular datasets, such as those from UCIrvine repo

See also the JAX codes listed at
https://github.com/janosh/awesome-normalizing-flows

**Comment by murphyk** [comment]
See  ch 24 of [vol 2](https://probml.github.io/pml-book/book2.html) for some details.

**Comment by vinaykakkad** [comment]
Hey, I would like to work on this issue.

**Comment by nsanghi** [comment]
I was trying to convert [jax-flow two moons demo](https://github.com/ChrisWaites/jax-flows/blob/master/examples/intro.ipynb) into a distrax/jax demo. Distrax does not have MADE and MAF bijectors. So was trying to use the ones from [tfp MADE](https://www.tensorflow.org/probability/api_docs/python/tfp/bijectors/AutoregressiveNetwork) and [tfp MAF](https://www.tensorflow.org/probability/api_docs/python/tfp/bijectors/MaskedAutoregressiveFlow) and convert them into `distrax bijectors` using `distraxt.as_bijectors` wrapper but getting different types of errors. Tried two approaches

```
from tensorflow_probability.substrates import jax as tfp
tfd = tfp.distributions
tfb = tfp.bijectors

made = tfb.AutoregressiveNetwork(params=2, hidden_units=[10, 10])
maf_bijector=tfb.MaskedAutoregressiveFlow(made)
maf_bijector_jax = distrax.as_bijector(maf_bijector)
```
This gives an error: `module 'tensorflow_probability.python.internal.backend.jax.numpy_keras' has no attribute 'constraints'` while executing `made = tfb.AutoregressiveNetwork(params=2, hidden_units=[10, 10])`

2nd approach, changed first import as below:

```
import tensorflow_probability as tfp
tfd = tfp.distributions
tfb = tfp.bijectors

made = tfb.AutoregressiveNetwork(params=2, hidden_units=[10, 10])
maf_bijector=tfb.MaskedAutoregressiveFlow(made)
maf_bijector_jax = distrax.as_bijector(maf_bijector)
```
I get an error for the line `maf_bijector_jax = distrax.as_bijector(maf_bijector)` saying `The numpy.ndarray conversion method __array__() was called on the JAX Tracer object Traced<ShapedArray`. Interestingly, `is_bijector` function `distrax/_src/utils/conversion.py in as_bijector` does not execute the path (that I expected it would)

```  
elif isinstance(obj, tfb.Bijector):
    return bijector_from_tfp.BijectorFromTFP(obj)
```
rather it branches to the path:

```
elif callable(obj):
    return lambda_bijector.Lambda(obj)
```

Is there any sample code or documentation on how to use tfp bijectors in distrax?

**Comment by vz415** [comment]
@nsanghi I don't think you'll be able to directly convert the MAF and MADE modules from `tfp`. Both inherit from `tfp.bijectors`, which is what I think the `distrax` documentation was referring to since the examples are mainly of simple bijectors, not entire architectures. I looked at how `tfp` [implements](https://github.com/tensorflow/probability/blob/v0.16.0/tensorflow_probability/python/bijectors/masked_autoregressive.py#L43-L387) the autoregressive property for both and they use a `tf.while` loop. If you want to build a custom flow in `distrax`, I think it may be easier to start with RealNVP, since that is composed of relatively more simple bijectors. Happy to help since I've been thinking about making the RealNVP in `distrax`.

**Comment by murphyk** [comment]
I think https://github.com/probml/probml-notebooks/pull/57 solves the issue, right?

**Comment by vz415** [comment]
Yes, it make a two moons example using a neural spline flow. I was suggesting how to approach making alternative normalizing flows in `distrax`.

**Comment by nsanghi** [comment]
@vz415 - thanks for the details. I was trying to implement autoregressive flows in distrax. So what you are saying is that [bijector_from_tfp](https://github.com/deepmind/distrax/blob/master/distrax/_src/bijectors/bijector_from_tfp.py) is not meant for complex bijectors. 
@murphyk - is there a plan to have autoregressive flows added to distrax. If not, I will follow tfp implementation and try to build a watered down one as part of an autoregressive flow example.

**Comment by vz415** [comment]
I believe so. [Here's](https://github.com/bayesiains/nflows) a helpful pytorch implementation of autoregressive flows from Iain Murray's lab to help make new flows in distrax.

**Comment by nsanghi** [comment]
Great. Thanks for the pointer. I will try to put some autoregressive flow in jax/distrax as a colab this weekend.
